Source/CBL_BlobStore.m

```diff
@@ -263,26 +323,122 @@ - (NSInteger) deleteBlobsExceptMatching: (BOOL(^)(CBLBlobKey))predicate
+        if (ok) {
+            [writer finish];
+            [writer install];
+        } else {
+            [writer cancel];
+            break;
+        }
+    }
+
+    if (ok) {
+        // Replace my directory with the new one:
+        Log(@"    Installing new blob store %@", tempPath);
+        if (CBLSafeReplaceDir(tempPath, _path, outError))
+            _encryptionKey = newKey;
+        else
+            ok = NO;
+    }
+    if (!ok) {
+        Warn(@"Changing blob-store encryption key failed! path= %@ ; error=%@",
+             _path, (outError ? *outError : nil));
+        [fmgr removeItemAtPath: tempPath error: NULL];
+    }
+    return ok;
+}
+
+
 - (NSString*) tempDir {
     if (!_tempDir) {
         // Find a temporary directory suitable for files that will be moved into the store:
+        _tempDir = [self createTempDir: NULL];
+        LogTo(CBLDatabase, @"CBL_BlobStore %@ created tempDir %@", _path, _tempDir);
+    }
+    return _tempDir;
+}
+
+
+- (NSString*) createTempDir: (NSError**)outError {
 #ifdef GNUSTEP
-        _tempDir = [NSTemporaryDirectory() copy];
+    NSString* name = $sprintf(@"CouchbaseLite-Temp-%@", CBLCreateUUID());
+    NSString* tempDir = [NSTemporaryDirectory() stringByAppendingPathComponent: name];
+    NSDictionary* attrs = @{NSFilePosixPermissions: @(0700)};
+    if (![[NSFileManager defaultManager] createDirectoryAtPath: tempDir
+                                   withIntermediateDirectories: YES
+                                                    attributes: attrs
+                                                         error: outError])
+        return nil;
+    return tempDir;
 #else
-        NSError* error;
-        NSURL* parentURL = [NSURL fileURLWithPath: _path isDirectory: YES];
-        NSURL* tempDirURL = [[NSFileManager defaultManager] 
-                                                 URLForDirectory: NSItemReplacementDirectory
-                                                 inDomain: NSUserDomainMask
-                                                 appropriateForURL: parentURL
-                                                 create: YES error: &error];
-        _tempDir = [tempDirURL.path copy];
-        Log(@"CBL_BlobStore %@ created tempDir %@", _path, _tempDir);
-        if (!_tempDir)
-            Warn(@"CBL_BlobStore: Unable to create temp dir: %@", error);
+    NSError* error;
+    NSURL* parentURL = [NSURL fileURLWithPath: _path isDirectory: YES];
+    NSURL* tempDirURL = [[NSFileManager defaultManager] URLForDirectory: NSItemReplacementDirectory
+                                                               inDomain: NSUserDomainMask
+                                                      appropriateForURL: parentURL
+                                                                 create: YES
+                                                                  error: outError];
+    if (!tempDirURL)
+        Warn(@"CBL_BlobStore: Unable to create temp dir: %@", error);
```

Comment:
outError might be NULL according to Obj-C calling conventions, so I added `error` so I could always report the error. But then I forgot to use `&error` in the calls above. I'll fix that.